File: pal/lib/src/parse.dart
// ignore_for_file: prefer_function_declarations_over_variables, constant_identifier_names, prefer_interpolation_to_compose_strings

import 'ast.dart';

typedef Token = (String, (int, int));
typedef Tokens = List<Token>;
typedef ParseCtx = (Tokens, int);
typedef Parser<T> = (T, ParseCtx) Function(ParseCtx);
typedef UnholeyParser<T> = (T, Tokens) Function(Tokens);
const PAREN_FOR = {true: '<', false: '('};
const MATCHING_PAREN = {'(': ')', '<': '>', '[': ']', '{': '}'};
const IN_WORD_CHARS = '_-';
const OUT_WORD_CHARS = ' \n<>()[]{},:.=';
const SPECIAL_CHARS = OUT_WORD_CHARS + IN_WORD_CHARS;
final HOLE_ID = RegExp(r'^_[0-9]*$');

Tokens tokenize(String s) {
  final ret = <Token>[];
  int line = 0;
  int col = 0;

  while (s.isNotEmpty) {
    var index = 0;
    if (!SPECIAL_CHARS.contains(s[index])) {
      index++;
      while (index < s.length && !OUT_WORD_CHARS.contains(s[index])) {
        index++;
      }
    }

    if (index == 0) {
      if (s[0] == '\n') {
        line++;
        col = 0;
      } else {
        if (s[0] != ' ') {
          ret.add((s.substring(0, 1), (line, col)));
        }
        col++;
      }
      s = s.substring(1);
    } else {
      ret.add((s.substring(0, index), (line, col)));
      col += index;
      s = s.substring(index);
    }
  }
  return ret;
}

final UnholeyParser<Program> parseProgram = (tokens) {
  if (tokens.isEmpty) {
    return ([[]], []);
  }
  if (tokens.first.$1 == '-') {
    final remaining = parseProgram(tokens.sublist(tokens.indexWhere((tok) => tok.$1 != '-')));
    return ([[], ...remaining.$1], remaining.$2);
  }
  assert(!SPECIAL_CHARS.contains(tokens[0].$1), tokens.toString());

  final id = tokens.first.$1;
  tokens = tokens.tail;
  Expr? type;
  if (tokens.first.$1 == ':') {
    tokens = tokens.tail;
    (type, tokens) = parseExpr(tokens);
  }
  Expr? value;
  if (tokens.first.$1 == '=') {
    tokens = tokens.tail;
    (value, tokens) = parseExpr(tokens);
  }

  final (remaining, []) = parseProgram(tokens);
  return (
    [
      [Binding(id, type, value), ...remaining.first],
      ...remaining.tail
    ],
    []
  );
};

Parser<void> _parseLit(String lit) => (ctx) {
      var (tokens, numHoles) = ctx;
      assert(tokens.isNotEmpty && tokens[0].$1 == lit, tokens.toString());
      return (null, (tokens.tail, numHoles));
    };

Parser<T> _then<T>(
  Parser<T> f1,
  Parser<void> f2,
) =>
    (tokens) {
      final (result, remaining) = f1(tokens);
      return (result, f2(remaining).$2);
    };

Parser<Expr> _parseFn(bool implicit) => (ctx) {
      String? id;
      Expr? maybe;
      Expr? argType;
      if (ctx case ([(':', _), ...var rest], var numHoles)) {
        id = null;
        (argType, ctx) = _parseExpr((rest, numHoles));
      } else {
        (maybe, ctx) = _parseExpr(ctx);
        final ([(token, (line, col)), ...rest], numHoles) = ctx;
        if ((maybe, token) case (Var(id: var varID), ':')) {
          ctx = (rest, numHoles);
          maybe = null;
          id = varID;
          var ([(token, _), ...], _) = ctx;
          if (token != ',' && token != MATCHING_PAREN[PAREN_FOR[implicit]]) {
            (argType, ctx) = _parseExpr(ctx);
          }
        } else if (token != ',' && token != MATCHING_PAREN[PAREN_FOR[implicit]]) {
          throw Exception('unexpected $token at $line:$col');
        }
      }
      final ([(_, pos), ...], _) = ctx;

      final ((result, next), kind) = switch (ctx) {
        ([(',', _), ...var remaining], var numHoles) => switch (
              _parseFn(implicit)((remaining, numHoles))) {
            (var fn, var next) => ((fn, next), (fn as Fn).kind)
          },
        ([(var e, _), (var n, (var line, var col)), ...var remaining], var numHoles)
            when e == MATCHING_PAREN[PAREN_FOR[implicit]] =>
          switch ((implicit, n)) {
            (true, '(') => switch (_parseFn(false)((remaining, numHoles))) {
                (var fn, var next) => ((fn, next), (fn as Fn).kind)
              },
            (_, '{') => (_then(_parseExpr, _parseLit('}'))((remaining, numHoles)), Fn.Def),
            (_, '[') => (_then(_parseExpr, _parseLit(']'))((remaining, numHoles)), Fn.Typ),
            _ => throw Exception('unexpected $n at $line:$col')
          },
        _ => throw Exception('unexpected $ctx')
      };
      ctx = next;

      if (maybe != null) {
        if (kind == Fn.Typ) {
          argType = maybe;
        } else {
          id = (maybe as Var).id;
        }
      }
      if (argType == null) {
        final (tokens, numHoles) = ctx;
        argType = Var('_$numHoles');
        ctx = (tokens, numHoles + 1);
      }
      return (Fn(implicit, kind, id, argType, result, t: pos), ctx);
    };

Parser<Expr> _parseFnAppBody(bool implicit, Expr fn) => (ctx) {
      var ([(_, pos), ...], _) = ctx;
      final (arg, remaining) = _parseExpr(ctx);
      final ([(tok, (line, col)), ...], _) = remaining;
      final end = MATCHING_PAREN[PAREN_FOR[implicit]]!;
      if (tok == end) {
        return _parseFnApp(App(implicit, fn, arg, t: pos))(
          _parseLit(end)(remaining).$2,
        );
      } else if (tok == ',') {
        return _parseFnAppBody(implicit, App(implicit, fn, arg, t: pos))(
          _parseLit(',')(remaining).$2,
        );
      } else {
        throw Exception('unexpected $tok at $line:$col');
      }
    };

Parser<Expr> _parseFnApp(Expr fn) => (ctx) {
      switch (ctx.$1) {
        case [('(', _), ...var rest]:
          final (expr, remaining) = _parseFnAppBody(false, fn)((rest, ctx.$2));
          return _parseFnApp(expr)(remaining);
        case [('<', _), ...var rest]:
          final (expr, remaining) = _parseFnAppBody(true, fn)((rest, ctx.$2));
          return _parseFnApp(expr)(remaining);
        default:
          return (fn, ctx);
      }
    };

final Parser<Expr> _parseExpr = (ctx) {
  var ([(token, pos), ...rest], numHoles) = ctx;

  if (token == '_') {
    token = '_$numHoles';
    numHoles++;
  }

  ctx = (rest, numHoles);

  final (initExpr, remaining) = switch (token) {
    '<' => _parseFn(true)(ctx),
    '(' => _parseFn(false)(ctx),
    _ => SPECIAL_CHARS.contains(token)
        ? throw Exception('unexpected $token')
        : (Var(token, t: pos), ctx),
  };

  return _parseFnApp(initExpr)(remaining);
};

final UnholeyParser<Expr> parseExpr = (tokens) {
  final (result, (remaining, _)) = _parseExpr((tokens, 0));
  return (result, remaining);
};

extension<T> on List<T> {
  List<T> get tail => sublist(1);
}

File: paljs/src/lang/parse.ts
import {Expr, Fn, FnKind, Program, Var, isVar, newVar} from './ast';

export type Pos = [number, number];
type Token = [string, Pos];
type ParseCtx = [Token[], number];
type Parser<T> = (_: ParseCtx) => [T, ParseCtx];
type UnholeyParser<T> = (_: Token[]) => [T, Token[]];

export const parenFor = (implicit: boolean | FnKind) => {
  if (implicit === 'type') return '[';
  if (implicit === 'def') return '{';
  if (implicit) return '<';
  return '(';
};

export const MATCHING_PAREN = {
  '(': ')',
  '<': '>',
  '[': ']',
  '{': '}',
} as const;

const IN_WORD_CHARS = '_-';
const OUT_WORD_CHARS = ' \n<>()[]{},:.=';
const SPECIAL_CHARS = OUT_WORD_CHARS + IN_WORD_CHARS;

export const tokenize = (s: string): Token[] => {
  const ret: Token[] = [];
  let line = 0;
  let col = 0;

  while (s.length !== 0) {
    let index = 0;
    if (!SPECIAL_CHARS.includes(s[index])) {
      index++;
      while (index < s.length && !OUT_WORD_CHARS.includes(s[index])) {
        index++;
      }
    }

    if (index === 0) {
      if (s[0] === '\n') {
        line++;
        col = 0;
      } else {
        if (s[0] !== ' ') {
          ret.push([s.substring(0, 1), [line, col]]);
        }
        col++;
      }
      s = s.substring(1);
    } else {
      ret.push([s.substring(0, index), [line, col]]);
      col += index;
      s = s.substring(index);
    }
  }
  return ret;
};

export const parseProgram: UnholeyParser<Program> = (tokens) => {
  if (tokens.length === 0) {
    return [[[]], []];
  }
  if (tokens[0][0] == '-') {
    const remaining = parseProgram(tokens.slice(tokens.findIndex((tok) => tok[0] != '-')));
    return [[[], ...remaining[0]], remaining[1]];
  }
  if (SPECIAL_CHARS.includes(tokens[0][0])) {
    throw new Error(tokens.toString());
  }

  const id = tokens[0][0];
  tokens = tokens.slice(1);
  let type: Expr | undefined;
  if (tokens[0][0] == ':') {
    tokens = tokens.slice(1);
    [type, tokens] = parseExpr(tokens);
  }
  let value: Expr | undefined;
  if (tokens[0][0] == '=') {
    tokens = tokens.slice(1);
    [value, tokens] = parseExpr(tokens);
  }

  const [remaining] = parseProgram(tokens);
  return [[[{id, type, value}, ...remaining[0]], ...remaining.slice(1)], []];
};

const parseLit =
  (lit: string): Parser<null> =>
  ([tokens, numHoles]) => {
    if (tokens.length === 0 || tokens[0][0] !== lit) {
      throw new Error(`${tokens[0][0]} != ${lit} at ${tokens[0][1]}`);
    }
    return [null, [tokens.slice(1), numHoles]];
  };

const then =
  <T>(f1: Parser<T>, f2: Parser<null>): Parser<T> =>
  (ctx) => {
    const [result, remaining] = f1(ctx);
    return [result, f2(remaining)[1]];
  };

const parseFn =
  (implicit: boolean): Parser<Fn> =>
  (ctx) => {
    let id: string | undefined;
    let maybe: Expr | undefined;
    let argType: Expr | undefined;
    const [[[tok]]] = ctx;
    if (tok === ':') {
      [argType, ctx] = _parseExpr(ctx);
    } else {
      [maybe, ctx] = _parseExpr(ctx);
      const [[[token, [line, col]], ...rest], numHoles] = ctx;
      if (isVar(maybe!) && token === ':') {
        ctx = [rest, numHoles];
        id = maybe.id;
        maybe = undefined;
        const [[[token]]] = ctx;
        if (token != ',' && token !== MATCHING_PAREN[parenFor(implicit)]) {
          [argType, ctx] = _parseExpr(ctx);
        }
      } else if (token != ',' && token != MATCHING_PAREN[parenFor(implicit)]) {
        throw new Error(`unexpected ${token} at ${line}:${col}`);
      }
    }
    const [[[, pos]]] = ctx;

    const [[result, next], kind] = ((): [[Expr, ParseCtx], 'def' | 'type'] => {
      const [[[tok]]] = ctx;
      let [[, ...remaining], numHoles] = ctx;
      if (tok === ',') {
        const [fn, next] = parseFn(implicit)([remaining, numHoles]);
        return [[fn, next], fn.kind];
      }
      if (tok !== MATCHING_PAREN[parenFor(implicit)]) {
        throw new Error(`unexpected ${ctx}`);
      }
      [[, ...remaining], numHoles] = ctx;
      ctx = [remaining, numHoles];
      const [[[nextParen, [line, col]]]] = ctx;
      [[, ...remaining]] = ctx;

      if (implicit && nextParen === '(') {
        const [fn, next] = parseFn(false)([remaining, numHoles]);
        return [[fn, next], fn.kind];
      } else if (nextParen === '{' || nextParen === '[') {
        const result = then(_parseExpr, parseLit(MATCHING_PAREN[nextParen]))([remaining, numHoles]);

        return [result, nextParen === '{' ? 'def' : 'type'];
      }
      throw new Error(`unexpected ${nextParen} at ${line}:${col}`);
    })();

    ctx = next;

    if (maybe !== undefined) {
      if (kind === 'type') {
        argType = maybe;
      } else {
        id = (maybe as Var).id;
      }
    }

    if (argType === undefined) {
      const [tokens, numHoles] = ctx;
      argType = newVar(`_${numHoles}`);
      ctx = [tokens, numHoles + 1];
    }
    return [{implicit, kind, argID: id, argType, result, meta: pos}, ctx];
  };

const parseFnAppBody =
  (implicit: boolean, fn: Expr): Parser<Expr> =>
  (ctx) => {
    const [[[, pos]]] = ctx;
    const [arg, remaining] = _parseExpr(ctx);
    const [[[tok, [line, col]]]] = remaining;
    const end = MATCHING_PAREN[parenFor(implicit)];
    if (tok == end) {
      return parseFnApp({kind: 'app', implicit, fn, arg, meta: pos})(parseLit(end)(remaining)[1]);
    } else if (tok == ',') {
      return parseFnAppBody(implicit, {
        kind: 'app',
        implicit,
        fn,
        arg,
        meta: pos,
      })(parseLit(',')(remaining)[1]);
    } else {
      throw new Error(`unexpected ${tok} at ${line}:${col}`);
    }
  };

const parseFnApp =
  (fn: Expr): Parser<Expr> =>
  (ctx) => {
    if (ctx[0].length === 0) return [fn, ctx];
    const [[[tok], ...rest], numHoles] = ctx;
    if (tok === '(' || tok === '<') {
      const [expr, remaining] = parseFnAppBody(tok === '<', fn)([rest, numHoles]);
      return parseFnApp(expr)(remaining);
    }
    return [fn, ctx];
  };

const _parseExpr: Parser<Expr> = (ctx) => {
  const [[[, pos], ...rest]] = ctx;
  let [[[token]], numHoles] = ctx;
  if (token == '_') {
    token = `_${numHoles}`;
    numHoles++;
  }

  ctx = [rest, numHoles];

  const [initExpr, remaining] = (() => {
    if (token === '<' || token === '(') {
      return parseFn(token === '<')(ctx);
    }
    if (SPECIAL_CHARS.includes(token)) {
      throw new Error(`unexpected ${token} at ${pos}`);
    }
    return [{kind: 'var', id: token, meta: pos} as const, ctx];
  })();

  return parseFnApp(initExpr)(remaining);
};

export const parseExpr: UnholeyParser<Expr> = (tokens) => {
  const [result, [remaining]] = _parseExpr([tokens, 0]);
  return [result, remaining];
};
